### PR TITLE
NEP: Update NEP 42 to note the issue of circular references

### PR DESCRIPTION
**Pull request by seberg**
This notes the issue a bit clearer and points out a possible
contribution, also list Marten as co-author, since he gave a large
chunk of input and discussion.

---

Mainly a minor update with respect to some of the array coercion discussion.

### CODE REVIEW

**Review by anirudh2290 — COMMENTED** [review]
Since you are on it, would like to point out that the "dtype hierarchy figure" link seems to be broken.

**Comment by seberg** [comment]
Thanks, should be fixed now... I think you mentioned it before...

File: doc/neps/nep-0042-new-dtypes.rst [hunk 1]
```diff
@@ -497,12 +498,23 @@ This shall be implemented by leveraging two pieces of information:
...
+A DType may choose to automatically discover from this Python type.
+This will be achieved using a global a mapping (dictionary-like) of::
+
+   known_python_types[type] = DType
+
+To anticipate the possibility of creating both a Python type (``pytype``)
+and ``DType`` dynamically, and thus the potential desire to delete them again,
+this mapping should generally be weak.
+This requires that the ``pytype`` holds on to the ``DType`` explicitly.
+Thus, additionally to building the global mapping, NumPy will store
```
Comment (rossbar) [inline doc/neps/nep-0042-new-dtypes.rst:512]:
```suggestion
Thus, in addition to building the global mapping, NumPy will store
```

File: doc/neps/nep-0042-new-dtypes.rst [hunk 2]
```diff
@@ -512,19 +524,20 @@ NumPy currently uses a small hard-coded mapping and conversion of numpy scalars
...
-    ``np.result_type(python_integer, array)``.
+    cached), so that they can provide the current value based promotion
+    functionality provided by ``np.result_type(python_integer, array)`` [1]_.
 
 To allow for a DType to accept specific inputs as known scalars, we will
-initially use a method such as ``known_scalar_type`` or ``known_scalar_object``
-which allows discovering an e.g. ``vector`` as a scalar (element) instead of
-a sequence (for the command ``np.array(vector, dtype=VectorDType)``).
+initially use a ``known_scalar_type`` method.
+This allows discovering a ``vector`` as a scalar (element) instead of
```
Comment (rossbar) [inline doc/neps/nep-0042-new-dtypes.rst:538]:
```suggestion
This allows discovery of a ``vector`` as a scalar (element) instead of
```

**Review by rossbar — APPROVED** [review]
LGTM, thanks @seberg !

**Comment by seberg** [comment]
Thanks Anirudh and Ross, its a small update, so I will just merge it.

**Pull request merged**